src/strawberry_sqlalchemy_mapper/field.py

```diff
@@ -0,0 +1,731 @@
+    sessionmaker: _SessionMaker | None = None,
+    keyset: Keyset | None = None,
+) -> Any:
+    ...
+
+
+@overload
+def connection(
+    graphql_type: type[relay.Connection[relay.NodeType]] | None = None,
+    *,
+    resolver: _RESOLVER_TYPE[NodeIterableType[Any]] | None = None,
+    name: str | None = None,
+    field_name: str | None = None,
+    is_subscription: bool = False,
+    description: str | None = None,
+    init: Literal[True] = True,
+    permission_classes: list[type[BasePermission]] | None = None,
+    deprecation_reason: str | None = None,
+    default: Any = dataclasses.MISSING,
+    default_factory: Callable[..., object] | object = dataclasses.MISSING,
+    metadata: Mapping[Any, Any] | None = None,
+    directives: Sequence[object] | None = (),
+    extensions: Sequence[FieldExtension] = (),
+    sessionmaker: _SessionMaker | None = None,
+    keyset: Keyset | None = None,
+) -> Any:
+    ...
+
+
+def connection(
+    graphql_type: type[relay.Connection[relay.NodeType]] | None = None,
+    *,
+    resolver: _RESOLVER_TYPE[NodeIterableType[Any]] | None = None,
+    name: str | None = None,
+    field_name: str | None = None,
+    is_subscription: bool = False,
+    description: str | None = None,
+    permission_classes: list[type[BasePermission]] | None = None,
+    deprecation_reason: str | None = None,
+    default: Any = dataclasses.MISSING,
+    default_factory: Callable[..., object] | object = dataclasses.MISSING,
+    metadata: Mapping[Any, Any] | None = None,
+    directives: Sequence[object] | None = (),
+    extensions: Sequence[FieldExtension] = (),
+    sessionmaker: _SessionMaker | None = None,
+    keyset: Keyset | None = None,
+    # This init parameter is used by pyright to determine whether this field
+    # is added in the constructor or not. It is not used to change
+    # any behavior at the moment.
+    init: Literal[True, False, None] = None,
+) -> Any:
+    """Annotate a property or a method to create a relay connection field.
+
+    Relay connections_ are mostly used for pagination purposes. This decorator
+    helps creating a complete relay endpoint that provides default arguments
+    and has a default implementation for the connection slicing.
+
+    Note that when setting a resolver to this field, it is expected for this
+    resolver to return an iterable of the expected node type, not the connection
+    itself. That iterable will then be paginated accordingly. So, the main use
+    case for this is to provide a filtered iterable of nodes by using some custom
+    filter arguments.
+
+    Examples
+    --------
+        Annotating something like this:
+
+        >>> @strawberry.type
+        >>> class X:
+        ...     some_node: relay.Connection[SomeType] = relay.connection(
+        ...         description="ABC",
+        ...     )
+        ...
+        ...     @relay.connection(description="ABC")
+        ...     def get_some_nodes(self, age: int) -> Iterable[SomeType]:
```

Comment:
How would someone do something like this were they extend the resolver to implement custom filtering without pulling all the data in from the database? Feels like either an abstraction is missing that would allow this to modify the query instead of returning results, or it would need to have access to the pagination params

Reply:
By returning a `Query` object. This is actually exactly what the `default_resolver` does when you don't define one, but you could do something like:

```python
@relay.connection(relay.ListConnection[SomeType])
def some_connection(self) -> Iterable[SomeType]:
    return session.query(SomeModel)
```

In this case `session.query(SomeModel)` will only access the database when being iterated, and what you return in here will be modified by the connection later: for `ListConnection` it will slice it, producing a `limit`/`offset` pagination. For the `KeysetConnection` I created it will do its thing

An example from strawberry-django: https://github.com/bellini666/strawberry-django-demo/blob/main/demo/order/schema.py#L28

Reply:
Amazing, thanks!